### PR TITLE
Chef Provisioner : Support real chef-zero/local mode

### PR DESCRIPTION
**Pull request by YpNo**
Support real chef-zero/local mode with version constraint

### CODE REVIEW

File: templates/provisioners/chef_zero/zero.erb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+<% end %>
+file_cache_path    "<%= file_cache_path %>"
+file_backup_path   "<%= file_backup_path %>"
+cookbook_path <%= cookbooks_path.inspect %>
+<% if roles_path %>
+role_path <%= roles_path.size == 1 ? roles_path.first.inspect : roles_path.inspect %>
+<% end %>
+log_level <%= log_level.inspect %>
+verbose_logging <%= verbose_logging.inspect %>
+<% if ! enable_reporting %>
```
Comment (sethvargo) [inline templates/provisioners/chef_zero/zero.erb]: We generally do not put an extra space between the bang and variable.

**Comment by sethvargo** [comment]
Hey @YpNo 

This is looking great! Thank you for the PR. Do you want to update the documentation for the Chef Zero provisioner to include and list these new options?

**Comment by YpNo** [comment]
Hi @sethvargo ,
I'll update the template and, yes I want to update the documentation for Chef Zero provisioner :)

**Comment by YpNo** [comment]
Done !

**Comment by sethvargo** [comment]
Looks good, thanks!

**Comment by YpNo** [comment]
Thanks too ! 
When the next release is scheduled?

**Comment by sethvargo** [comment]
We do not have a release scheduled at this time.

**Comment by bdwyertech** [comment]
@YpNo Why did you pull ```nodes_path``` out?

**Comment by sethvargo** [comment]
@bdwyertech this has been completely refactored in Vagrant 1.7.3

**Comment by bdwyertech** [comment]
Yeah, noticed that @sethvargo . So, it seems we can no longer feed a ```nodes_path``` in to facilitate search and such.

**Comment by sethvargo** [comment]
@bdwyertech please open a new issue. Thanks!

**Comment by drags** [comment]
Is this provisioner still working within Vagrant 1.7.x?

Judging by the code and some recent executions this seems to run something like `cd /tmp/vagrant-chef/<some-dir> && chef-zero` which would require all of the data (cookbooks/roles/data_bags) to be stored together in that directory. However when the chef-zero provisioner mounts the directories it puts them in separate folders (`/tmp/vagrant-chef/<hash>/<mounted-folder>`).

The only way around this that I've seen is to use `knife <subcommand> -z`  to upload the data into the `chef-zero` server, but I can't find any place in the chef_zero provisioner code where this happens.

Can you confirm?

Running Vagrant 1.7.4
Plugin list:
```$ vagrant plugin list
vagrant-share (1.1.4, system)```

**Comment by sethvargo** [comment]
Hi @drags 

Are you experiencing issues with the provisioner? We tested this before merging and it works just fine. There is an issue with nodes_path that has since been fixed since release.

**Comment by drags** [comment]
Hi @sethvargo,

I am experiencing issues with the provisioner, here's a paste of the (non debug) log output to give some context:

```
$ vagrant destroy -f; time vagrant up --provision
==> default: VM not created. Moving on...
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'ubuntu_nocm'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: chef-repo_default_1439861960499_8145
==> default: Fixed port collision for 22 => 2222. Now on port 2200.
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 => 2200 (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2200
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Connection timeout. Retrying...
    ==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Mounting shared folders...
    default: /vagrant => /Users/tim/src/chef-repo
    default: /tmp/vagrant-chef/8addf10b12dfed40403da41012b18717/roles => /Users/tim/src/chef-repo/roles
    default: /tmp/vagrant-chef/f9ae443f73e3f920af0c78cbbb967ef4/cookbooks => /Users/tim/src/chef-repo/dep_cookbooks
    default: /tmp/vagrant-chef/6fd3da73972d018f85ba5733c51be7a2/cookbooks => /Users/tim/src/chef-repo/cookbooks
    default: /tmp/vagrant-chef/f5904b8c9b12a4eb12e865c109951859/data_bags => /Users/tim/src/chef-repo/data_bags
==> default: Running provisioner: chef_zero...
    default: Installing Chef (latest)...
Generating chef JSON and uploading...
==> default: Running chef-client (local-mode)...
==> default: stdin: is not a tty
==> default: [2015-08-18T01:40:23+00:00] INFO: Started chef-zero at chefzero://localhost:8889 with repository at /tmp/vagrant-chef/6fd3da73972d018f85ba5733c51be7a2, /tmp/vagrant-chef/f9ae443f73e3f920af0c78cbbb967ef4
==> default:   One version per cookbook
==> default:   data_bags at /tmp/vagrant-chef/f5904b8c9b12a4eb12e865c109951859/data_bags
==> default:   roles at /tmp/vagrant-chef/8addf10b12dfed40403da41012b18717/roles
==> default: [2015-08-18T01:40:23+00:00] INFO: Forking chef instance to converge...
==> default: Starting Chef Client, version 12.4.1
==> default: [2015-08-18T01:40:23+00:00] INFO: *** Chef 12.4.1 ***
==> default: [2015-08-18T01:40:23+00:00] INFO: Chef-client pid: 1808
==> default: [2015-08-18T01:40:29+00:00] INFO: HTTP Request Returned 404 Not Found: Object not found: chefzero://localhost:8889/nodes/vagrant-ubuntu-trusty-64
==> default: [2015-08-18T01:40:29+00:00] INFO: Setting the run_list to ["role[basev2]"] from CLI options
==> default: [2015-08-18T01:40:29+00:00] INFO: HTTP Request Returned 404 Not Found: Object not found: chefzero://localhost:8889/roles/basev2
==> default: [2015-08-18T01:40:29+00:00] ERROR: Role basev2 (included by 'top level') is in the runlist but does not exist. Skipping expand.
==> default:
==> default: ================================================================================
==> default: Error expanding the run_list:
==> default: ================================================================================
==> default:
==> default:
==> default: Missing Role(s) in Run List:
==> default: ----------------------------
==> default: * basev2 included by 'top level'
==> default:
==> default:
==> default: Original Run List
==> default: -----------------
==> default: * role[basev2]
==> default:
==> default:
==> default: Running handlers:
==> default: [2015-08-18T01:40:29+00:00] ERROR: Running exception handlers
==> default: Running handlers complete
==> default: [2015-08-18T01:40:29+00:00] ERROR: Exception handlers complete
==> default: Chef Client failed. 0 resources updated in 6.134546019 seconds
==> default: [2015-08-18T01:40:29+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
==> default: [2015-08-18T01:40:29+00:00] ERROR: The expanded run list includes nonexistent roles: basev2
==> default: [2015-08-18T01:40:31+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
Chef never successfully completed! Any errors should be visible in the
output above. Please fix your recipes so that they properly complete.

real    1m20.281s
user    0m4.126s
sys     0m1.749s
```

If I attempt to run `curl -H "Accept: application/json" localhost:8889/roles` on the box during the `chef-client -z` run then I get this output:
```
{

}
```

If I attempt to run `curl -H "Accept: application/json" localhost:8889/cookbooks` during the `chef-client -z` run then I get a 0 length response.

I was trying to debug from the chef-zero side, and I'm unclear on how this is supposed to work (especially with the `chef_zero` provisioner accepting an array of cookbook paths). From what I can tell chef-zero will allow either 1) serving from a directory which it expects to be a server repo directory (with `./roles`. `./cookbooks`, etc), or 2) act as a local listener where `knife <subcommand> -z` is used to upload items to the `chef-zero` server.

For reference I've put a full `vagrant up --provision --debug` output here: https://gist.github.com/drags/d6d4c0a7e8b3e13b8bca (oh goodness, I almost posted my AWS keys since vagrant debug output includes full `env`.. that is.. not a safe default)

I've tried both with latest chef as well as 11.18

My Vagrantfile: 
```
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu_nocm"

  config.vm.provider "virtualbox" do |vb|
     vb.customize ["modifyvm", :id, "--memory", "4096"]
  end

  config.vm.provision "chef_zero" do |chef|
    chef.cookbooks_path = ["cookbooks", "dep_cookbooks"]
    chef.data_bags_path = "data_bags"
    chef.roles_path = "roles"
    chef.add_role "basev2"

  end

end
```

I'm running into roadblocks on all fronts trying to debug, all pointers/suggestions greatly appreciated.

Thanks,
Tim

**Comment by sethvargo** [comment]
Hi @drags 

This is a closed and merged PR. There is no way to track this work or re-open it for triage. Can you please open a new issue so we can tag appropriately? Thanks!

**Pull request merged**